### ISSUE TITLE
micropip does not resolve locally available versions 

### ISSUE THREAD

**Issue by freakboy3742** [pyodide/pyodide#2408]
I have three wheels:

* travertino is v0.1.3
* toga-core is v0.3.0.dev33, and declares travertino>=0.1.3 as a requirement
* toga-web is v0.3.0.dev33, and declares toga-core==0.3.0.dev33 as a requirement

All three exist as `.whl` files in my static files folder. Travertino 0.1.3 has been published to PyPI; toga-core and toga-web dev32 have been published to PyPI, but dev33 are local development builds. 

On my Pyodide page, I run:
```
import micropip
micropip.install([
    './static/travertino-0.1.3-py3-none-any.whl',
    './static/toga_core-0.3.0.dev33-py3-none-any.whl',
    './static/toga_web-0.3.0.dev33-py3-none-any.whl',
])
```
This raises the following exception:
```
Traceback (most recent call last):
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/lib/python3.10/asyncio/tasks.py", line 234, in __step
    result = coro.throw(exc)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 183, in install
    transaction = await self.gather_requirements(requirements, ctx, keep_going)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 173, in gather_requirements
    await gather(*requirement_promises)
  File "/lib/python3.10/asyncio/futures.py", line 284, in __await__
    yield self  # This tells Task to wait for completion.
  File "/lib/python3.10/asyncio/tasks.py", line 304, in __wakeup
    future.result()
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/lib/python3.10/asyncio/tasks.py", line 232, in __step
    result = coro.send(None)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 245, in add_requirement
    await self.add_wheel(name, wheel, version, (), ctx, transaction)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 316, in add_wheel
    await self.add_requirement(recurs_req, ctx, transaction)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 286, in add_requirement
    raise ValueError(
ValueError: Couldn't find a pure Python 3 wheel for 'toga-core==0.3.0.dev33'. You can use `micropip.install(..., keep_going=True)` to get a list of all packages with missing wheels.
```

If I relax the `install_requires` on toga-web to remove the declared dependency on toga-core, the import succeeds, and the code runs successfully.

I've tried splitting the install into two passes:
```
await micropip.install([
    './static/travertino-0.1.3-py3-none-any.whl',
    './static/toga_core-0.3.0.dev33-py3-none-any.whl',
])
await micropip.install([
    './static/toga_web-0.3.0.dev33-py3-none-any.whl',
])
```
thinking that if toga-core's installation was completed, toga-web would be able to find it; that makes no difference. 

I've also tried producing the packages without the .dev33 version suffix, thinking that it might be "prerelease" versions that are the issue, but that made no difference either.

However, if I revert the version number on toga-core and toga-web to 0.3.0.dev32 (the last versions published to PyPI), the import *succeeds* - but the old version (published to PyPI) version of the toga-core code is used. 

Looking at the network requests that the page has made indicates that the wheels were all loaded - and then micropip requested the toga-core PyPI metadata, and toga_core-0.3.0.dev32 (the most recent published version) was downloaded from PyPI. Interestingly, *no* request was made to load Travertino 0.1.3 from PyPI as a result of the dependency declared in toga-core.

**Comment by ryanking13** [comment]
Thanks for the report. Yes I can confirm that this a bug and splitting the install should work.

The reason of this bug is that micropip is handling the package name differently when it is installed from PyPI or from `.whl` url.

```python
>>> import micropip
>>> await micropip.install("toga-web==0.3.0.dev32")
>>> micropip.list()
Name       | Version     | Source
---------- | ----------- | -------
travertino | 0.1.3       | pypi
toga-core  | 0.3.0.dev32 | pypi
toga-web   | 0.3.0.dev32 | pypi
...
```

```python
>>> import micropip
>>> await micropip.install("https://files.pythonhosted.org/packages/56/93/7bf199546e0de51404d7c714c618f9e67d215c2723351e1af0392e5e0a59/toga_web-0.3.0.dev32-py3-none-any.whl")
>>> micropip.list()
Name       | Version     | Source
---------- | ----------- | ------------------------------------------------------------------------------------------------------------------------------------------------
travertino | 0.1.3       | pypi
toga-core  | 0.3.0.dev32 | pypi
toga_web   | 0.3.0.dev32 | https://files.pythonhosted.org/packages/56/93/7bf199546e0de51404d7c714c618f9e67d215c2723351e1af0392e5e0a59/toga_web-0.3.0.dev32-py3-none-any.whl
...
```

When installed from PyPI, micropip think its name is `toga-web`, while when installed from whl url, micropip parses the name from whl url: `toga_web`.

**Comment by rth** [comment]
Thanks for the report. I think in any case  it would be good to be able to disable dependency resolution to force install a set of packages https://github.com/pyodide/pyodide/issues/2412 

> The reason of this bug is that micropip is handling the package name differently when it is installed from PyPI or from .whl URL.

Indeed, as far as I remember this was done because for PyPI packages we have the package index with all the metadata, while for custom URL everything is inside the wheel. So far we are parsing the wheel name, to determine if it's a pure python wheel and also how to reference the package in question by name. The issue is that all  `-` in the package name get transformed into `_` in the wheel filename. 

I think it should work if we take the package name from inside the wheel metadata [here](https://github.com/pyodide/pyodide/blob/main/packages/micropip/src/micropip/_micropip.py#L314) and correct it if it doesn't match. Though then we would need to make sure the names  in `transaction["locked"]` also use this transformation with `_` since that is set before the wheel is downloaded.

**Comment by freakboy3742** [comment]
@rth Makes sense; and I agree that `no_deps` would be a good workaround - possibly even a preferable one for "production" situations where you'd be looking for an analog of lockfile handling so you know exactly what is being delivered to the client.

**Comment by ryanking13** [comment]
Closing as resolved via #2414. If you still have an issue, please re-open this issue.
